include/nbl/builtin/hlsl/emulated/emulated_float64_t_utils.hlsl

```diff
@@ -0,0 +1,459 @@
+
+    EmulatedType x;
+    EmulatedType y;
+    EmulatedType z;
+
+    EmulatedType calcComponentSum() NBL_CONST_MEMBER_FUNC
+    {
+        return x + y + z;
+    }
+
+    type operator*(NBL_CONST_REF_ARG(type) rhs) NBL_CONST_MEMBER_FUNC
+    {
+        type output;
+        output.x = x * rhs.x;
+        output.y = y * rhs.y;
+        output.z = z * rhs.z;
+
+        return output;
+    }
+};
+
+template<typename EmulatedType>
+struct emulated_vector<EmulatedType, 4>
+{
+    using type = emulated_vector<EmulatedType, 4>;
+
+    EmulatedType x;
+    EmulatedType y;
+    EmulatedType z;
+    EmulatedType w;
+};
+
+template<typename EmulatedType>
+using emulated_vector_t2 = emulated_vector<EmulatedType, 2>;
+template<typename EmulatedType>
+using emulated_vector_t3 = emulated_vector<EmulatedType, 3>;
+template<typename EmulatedType>
+using emulated_vector_t4 = emulated_vector<EmulatedType, 4>;
+
+namespace impl
+{
+
+template<typename T, typename U, uint32_t Dim>
+struct static_cast_helper<vector<T,Dim>,emulated_vector<U,Dim>,void>
+{
+    static inline vector<T,Dim> cast(emulated_vector<U,Dim> vec)
+    {
+        return vector<T,Dim>(_static_cast<T,U>(vec.x), _static_cast<T,U>(vec.y));
+    }
+};
+
+}
+
+//template<typename EmulatedType, uint32_t N, uint32_t M>
+//struct emulated_matrix_base
+//{
+//    using vec_t = emulated_vector<EmulatedType, N>;
+//    vec_t columns[M];
+//};
+
+template<typename EmulatedType, uint32_t N, uint32_t M>
+struct emulated_matrix {}; // : emulated_matrix_base<EmulatedType, N, M> {};
+
+template<typename EmulatedType>
+struct emulated_matrix<EmulatedType, 2, 2>// : emulated_matrix_base<EmulatedType, 2, 2>
+{
+    using vec_t = emulated_vector_t2<EmulatedType>;
+    using type = emulated_matrix<EmulatedType, 2, 2>;
+
+    vec_t columns[2];
+
+    type getTransposed() NBL_CONST_MEMBER_FUNC
+    {
+        type output;
+
+        output.columns[0].x = columns[0].x;
+        output.columns[1].x = columns[0].y;
+
+        output.columns[0].y = columns[1].x;
+        output.columns[1].y = columns[1].y;
+
+        return output;
+    }
+
+    type operator*(NBL_CONST_REF_ARG(type) rhs) NBL_CONST_MEMBER_FUNC
+    {
+        type output;
+        type lhsTransposed = getTransposed();
+
+        output.columns[0].x = (lhsTransposed.columns[0] * rhs.columns[0]).calcComponentSum();
+        output.columns[0].y = (lhsTransposed.columns[0] * rhs.columns[1]).calcComponentSum();
+
+        output.columns[1].x = (lhsTransposed.columns[1] * rhs.columns[0]).calcComponentSum();
+        output.columns[1].y = (lhsTransposed.columns[1] * rhs.columns[1]).calcComponentSum();
+
+        return output.getTransposed();
+    }
+
+    vec_t operator*(NBL_CONST_REF_ARG(vec_t) rhs)
+    {
+        vec_t output;
+        type lhsTransposed = getTransposed();
+
+        output.x = (columns[0] * rhs).calcComponentSum();
+        output.y = (columns[1] * rhs).calcComponentSum();
+
+        return output;
+    }
```

Comment:
in HLSL the `operator*` multiplies matrices point (value) wise (so its only ok for matrix-scalar mul), to get a real matrix-matrix or matrix-vector mul, there's a `mul()` global function